### ISSUE TITLE
Can no longer install with Doctrine DBAL 2.5

### ISSUE THREAD

**Issue by leofeyer** [contao/installation-bundle#54]
Doctrine DBAL 2.5 has added the [`server_version`](http://symfony.com/doc/current/reference/configuration/doctrine.html#doctrine-dbal-configuration) parameter. It it is not given, Doctrine tries to read it from the MySQL server, thus enabling a database connection, which leads to an `Access denied for user ''@'localhost'` exception.

![](https://cloud.githubusercontent.com/assets/1192057/26574240/e07ff0d8-4520-11e7-9be7-c156d0cf6229.png)

Related issue: https://github.com/kriswallsmith/assetic/issues/681

**Comment by leofeyer** [comment]
@contao/developers @fritzmg /cc

**Comment by aschempp** [comment]
> Doctrine tries to read it from the MySQL server

Who's Doctrine? The bundle? The bundle extension?

**Comment by Toflar** [comment]
When does that error happen? When you call the install tool? The screenshot does not include the whole stack trace :)

**Comment by leofeyer** [comment]
The problem is not limited to the install tool. It occurs everywhere.

<img width="1046" alt="" src="https://cloud.githubusercontent.com/assets/1192057/26577451/b7321f42-452c-11e7-9282-025c5dad0bcf.png">

**Comment by leofeyer** [comment]
See: https://github.com/doctrine/dbal/issues/990
See: https://github.com/doctrine/DoctrineBundle/issues/351

**Comment by aschempp** [comment]
The quickest solution would be to add a conflict with 2.5?

**Comment by aschempp** [comment]
The solution is fairly simple. We cannot assume a complete installation without a working DB connection. Adjust to the following:

```php
// class Contao\Config
public static function isComplete()
{
	if (static::$blnHasLcf === null || !static::has('licenseAccepted'))
	{
		return false;
	}

	try
	{
		\Database::getInstance()->listTables();
	}
	catch (\Exception $e)
	{
		return false;
	}

	return true;
}
```

**Comment by fritzmg** [comment]
@aschempp but this would introduce two additional database queries on each request.

**Comment by aschempp** [comment]
True, we should probably perform the DB check only when we're in the install tool.

Another option would be to store `isComplete` in the container, but that's rather strange imho.

**Comment by leofeyer** [comment]
This does not fix the issue, because the problem originates somewhere else.

https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Connection.php#L429

As you can see in the stack trace, the `detectDatabasePlatform()` method is causing the issue, because it tries to connect to the database before the database credentials have been set.

**Comment by aschempp** [comment]
It does fix the issue in my installation. `detectDatabasePlatform` is only called because someone tries to connect, in my case it's the cron job listener checking if table `tl_cron` exists.

**Comment by leofeyer** [comment]
Nope, sorry. Your fix is not working.

**Comment by leofeyer** [comment]
And just so you see that the issue is **not at all** related to the install tool:

<img width="1242" alt="" src="https://cloud.githubusercontent.com/assets/1192057/26582917/2f0ae01a-4543-11e7-911f-f2f568c2e730.png">

**Comment by aschempp** [comment]
Well to me the install tool must work without DB, everything else cannot work. Idk why the command tries to connect to the database?

**Comment by leofeyer** [comment]
Because Doctrine DBAL tries to connect to the database to retrieve the version number.

https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Connection.php#L429

And this already happens in `ConnectionFactory::createConnection()`, which is executed upon `$container->get('doctrine.dbal.default_connection');`.

And all of this happens way before the installation controller is even constructed!

**Comment by aschempp** [comment]
In my case no connection to the database was made if noone queried something…

**Comment by fritzmg** [comment]
This error will probably only occur, if you do not allow anonymous connections to your database server (see also https://github.com/contao/core-bundle/issues/736).

**Comment by leofeyer** [comment]
Yes, that's true. If anonymous connections are allowed, the issue does not occur. But that's rarely the case fortunately.

**Comment by leofeyer** [comment]
I have commented [here](https://github.com/doctrine/dbal/pull/2671#issuecomment-304875595) to document the issue.

**Comment by leofeyer** [comment]
BTW, this issue also affects the managed edition:

<img width="1242" alt="" src="https://cloud.githubusercontent.com/assets/1192057/26586696/4b7478f8-4550-11e7-9591-c212c3c5dc97.png">

**Comment by leofeyer** [comment]
It seems that our [`BinaryStringType`](https://github.com/contao/standard-edition/blob/release/4.4.0/app/config/config.yml#L47-L50) is the reason why the `getDatabasePlatform()` method is called in the [`ConnectionFactory`](https://github.com/doctrine/DoctrineBundle/blob/master/ConnectionFactory.php#L65-L70) class. We might be able to register it in an compiler path to work around it.

**Comment by mattjanssen** [comment]
Sounds like your GUI config bundle needs to be able to handle when the DB credentials are missing or are no longer valid? Perhaps you could register a `kernel.exception` event listener that disables the `doctrine` config section once a DB `ConnectionException` is thrown. It could do this by either editing config.yml or switching to the "install" environment which would load config_install.yml. Finally it redirects the user with a `RedirectResponse` to the GUI admin login page with a nice error.

This is assuming your GUI config bundle is in the business of editing config.yml, which seems like a possibility :)

**Comment by leofeyer** [comment]
@mattjanssen Thanks a lot for your input. Much appreciated.

**Comment by leofeyer** [comment]
We are now dynamically adding `server_version: 5.1` in the Contao managed edition, if there is no database connection: https://github.com/contao/manager-bundle/commit/06f1f294bedcbe456869145597de0dfa32f352aa

**Comment by aschempp** [comment]
@leofeyer should we add an `is_array` check in https://github.com/contao/manager-bundle/commit/fd465bce6d55491c652dfad49a695485b6dd69b1#diff-e3375107bffc209723fa127bb186d8c6R146 ?

**Comment by leofeyer** [comment]
If `$extensionConfig['dbal']['connections']['default']` is set, it should always be an array, shouldn't it?

**Comment by Toflar** [comment]
Unit tests would tell you :P

**Comment by leofeyer** [comment]
No true in this case. If `$extensionConfig['dbal']['connections']['default']` is not an array, the user has misconfigured the application.

```yml
doctrine:
    dbal:
        connections:
            default: 'string'  <--  this is wrong!
```

**Comment by aschempp** [comment]
> Unit tests would tell you :P

Not really, because the data is coming from other bundles. I think the unifying of TreeBuilder is only done after our method. According to https://github.com/doctrine/DoctrineBundle/blob/master/DependencyInjection/Configuration.php#L69 it's actually possible to set the connection details directly in `doctrine.dbal` but I think that's just an unsupported case for us then, especially because we know that the config loaded by manager-bundle will not be like that.

**Comment by fritzmg** [comment]
Is this actually fixed in Contao 4.4.0? I am unable to open the Install Tool once I have removed the anonymous user accounts from my SQL database. Stack trace seems to be the same as posted by @leofeyer 

// nvm, suddenly it worked after clearing the dev cache
